sys/suit/coap.c

```diff
@@ -0,0 +1,513 @@
+    if (res < 0) {
+        return res;
+    }
+
+
+    int more = 1;
+    size_t num = 0;
+    res = -1;
+    while (more == 1) {
+        DEBUG("fetching block %u\n", (unsigned)num);
+        res = _fetch_block(&pkt, buf, &sock, path, blksize, num);
+        DEBUG("res=%i\n", res);
+
+        if (!res) {
+            coap_block1_t block2;
+            coap_get_block2(&pkt, &block2);
+            more = block2.more;
+
+            if (callback(arg, block2.offset, pkt.payload, pkt.payload_len, more)) {
+                DEBUG("callback res != 0, aborting.\n");
+                res = -1;
+                goto out;
+            }
+        }
+        else {
+            DEBUG("error fetching block\n");
+            res = -1;
+            goto out;
+        }
+
+        num += 1;
+    }
+
+out:
+    sock_udp_close(&sock);
+    return res;
+}
+
+int suit_coap_get_blockwise_url(const char *url,
+                               coap_blksize_t blksize,
+                               coap_blockwise_cb_t callback, void *arg)
+{
+    char hostport[SOCK_HOSTPORT_MAXLEN];
+    char urlpath[SOCK_URLPATH_MAXLEN];
+    sock_udp_ep_t remote;
+
+    if (strncmp(url, "coap://", 7)) {
+        LOG_INFO("suit: URL doesn't start with \"coap://\"\n");
+        return -EINVAL;
+    }
+
+    if (sock_urlsplit(url, hostport, urlpath) < 0) {
+        LOG_INFO("suit: invalid URL\n");
+        return -EINVAL;
+    }
+
+    if (sock_udp_str2ep(&remote, hostport) < 0) {
+        LOG_INFO("suit: invalid URL\n");
+        return -EINVAL;
+    }
+
+    if (!remote.port) {
+        remote.port = COAP_PORT;
+    }
+
+    return suit_coap_get_blockwise(&remote, urlpath, blksize, callback, arg);
+}
+
+typedef struct {
+    size_t offset;
+    uint8_t *ptr;
+    size_t len;
+} _buf_t;
+
+static int _2buf(void *arg, size_t offset, uint8_t *buf, size_t len, int more)
+{
+    (void)more;
+
+    _buf_t *_buf = arg;
+    if (_buf->offset != offset) {
+        return 0;
+    }
+    if (len > _buf->len) {
+        return -1;
+    }
+    else {
+        memcpy(_buf->ptr, buf, len);
+        _buf->offset += len;
+        _buf->ptr += len;
+        _buf->len -= len;
+        return 0;
+    }
+}
+
+ssize_t suit_coap_get_blockwise_url_buf(const char *url,
+                               coap_blksize_t blksize,
+                               uint8_t *buf, size_t len)
+{
+    _buf_t _buf = { .ptr=buf, .len=len };
+    int res = suit_coap_get_blockwise_url(url, blksize, _2buf, &_buf);
+    return (res < 0) ? (ssize_t)res : (ssize_t)_buf.offset;
+}
+
+static void _suit_handle_url(const char *url)
+{
+    LOG_INFO("suit_coap: downloading \"%s\"\n", url);
+    ssize_t size = suit_coap_get_blockwise_url_buf(url, COAP_BLOCKSIZE_64, _manifest_buf,
+                                              SUIT_MANIFEST_BUFSIZE);
+    if (size >= 0) {
+        LOG_INFO("suit_coap: got manifest with size %u\n", (unsigned)size);
+
+        riotboot_flashwrite_t writer;
```

Comment:
This contains a `uint8_t flashpage_buf[FLASHPAGE_SIZE];` with `FLASHPAGE_SIZE` being up to 8k on some platforms - is it a good idea to put that on the stack?